src/ast/analysis/Type.cpp

```diff
@@ -823,23 +840,282 @@ void TypeAnalysis::print(std::ostream& os) const {
+}
+
 void TypeAnalysis::run(const TranslationUnit& translationUnit) {
     // Check if debugging information is being generated
     std::ostream* debugStream = nullptr;
     if (Global::config().has("debug-report") || Global::config().has("show", "type-analysis")) {
         debugStream = &analysisLogs;
     }
 
-    // Analyse types, clause by clause.
-    const Program& program = translationUnit.getProgram();
-    for (const Clause* clause : program.getClauses()) {
-        auto clauseArgumentTypes = analyseTypes(translationUnit, *clause, debugStream);
-        argumentTypes.insert(clauseArgumentTypes.begin(), clauseArgumentTypes.end());
+    // Utility lambdas to determine if all args are of the same type.
+    auto isFloat = [&](const Argument* argument) {
+        return isOfKind(getTypes(argument), TypeAttribute::Float);
+    };
+    auto isUnsigned = [&](const Argument* argument) {
+        return isOfKind(getTypes(argument), TypeAttribute::Unsigned);
+    };
+    auto isSymbol = [&](const Argument* argument) {
+        return isOfKind(getTypes(argument), TypeAttribute::Symbol);
+    };
 
-        if (debugStream != nullptr) {
-            // Store an annotated clause for printing purposes
-            annotatedClauses.emplace_back(createAnnotatedClause(clause, clauseArgumentTypes));
+    // Analyse user-defined functor types
+    const Program& program = translationUnit.getProgram();
+    visitDepthFirst(
+            program, [&](const FunctorDeclaration& fdecl) { udfDeclaration[fdecl.getName()] = &fdecl; });
+
+    // Rest of the analysis done until fixpoint reached
+    bool changed = true;
+    while (changed) {
+        changed = false;
+        argumentTypes.clear();
+
+        // Analyse general argument types, clause by clause.
+        for (const Clause* clause : program.getClauses()) {
+            auto clauseArgumentTypes = analyseTypes(translationUnit, *clause, debugStream);
+            argumentTypes.insert(clauseArgumentTypes.begin(), clauseArgumentTypes.end());
+
+            if (debugStream != nullptr) {
+                // Store an annotated clause for printing purposes
+                annotatedClauses.emplace_back(createAnnotatedClause(clause, clauseArgumentTypes));
+            }
         }
+
+        // Analyse intrinsic-functor types
+        visitDepthFirst(program, [&](const IntrinsicFunctor& functor) {
+            auto candidates = validOverloads(functor);
+            if (candidates.empty()) {
+                // No valid overloads - mark it as an invalid functor
+                if (contains(functorInfo, &functor)) {
+                    functorInfo.erase(&functor);
+                    changed = true;
+                }
+                return;
+            }
+
+            // Update to the canonic representation if different
+            const auto* curInfo = &candidates.front().get();
+            if (contains(functorInfo, &functor) && functorInfo.at(&functor) == curInfo) return;
```

Comment:
Single line if blocks like this hide the contents and do not make for easily readable code. If you must use them where they affect control flow like this, at least use a blank line so the change can be seen